flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -211,7 +215,8 @@ private Runnable restartTasks(final Set<ExecutionVertexVersion> executionVertexV
 	}
 
 	private CompletableFuture<?> cancelExecutionVertex(final ExecutionVertexID executionVertexId) {
-		return executionVertexOperations.cancel(getExecutionVertex(executionVertexId));
+		return executionVertexOperations.cancel(getExecutionVertex(executionVertexId))
+			.whenComplete((Object ignored, Throwable t) -> executionSlotAllocator.cancel(executionVertexId));
```

Comment:
It's similar to the slot request cancellation registered in `Execution#allocateAndAssignSlotForExecution`.

In my understanding, doing it after the cancellation finishes means the vertex(Execution) has reached a stable terminated state then. So that the later calls triggered by executionSlotAllocator.cancel() can return early without any complex logic that may change the vertex state. This can make things easier.

Reply:
In the current PR, can it happen that we cancel the slot allocation of a newer deployment? That is, can it happen that while we wait for the cancel future to complete, we already request a new slot for one of the vertices that are being canceled. Currently there can be only one ongoing allocation per `ExecutionVertex` (see `DefaultExecutionSlotAllocator#pendingSlotAssignments`). Therefore, I thought it's safest to cancel potential ongoing allocations in `allocateSlotsAndDeploy()`, i.e., cancel slots before allocating new slots.

Reply:
It should not happen because  the `cancelFuture` returned by `cancelExecutionVertex` completes only if `executionSlotAllocator.cancel()` completes.
And a vertex will only be restarted if its `cancelFuture` completes.

We can add a check in `allocateSlotsAndDeploy()` to guarantee that `DefaultExecutionSlotAllocator` has no pending slot assignment for the vertices to schedule. The check will fail only if there is a bug.

Reply:
We cannot get the slot assignment from ExecutionSlotAllocator in scheduler. 

Maybe we should do the check in DefaultExecutionSlotAllocator#allocateSlotsFor(...) to ensure that the vertex has no prior pending slot request? 
I think this change can also be beneficial for DefaultExecutionSlotAllocator so that it will not drop reference of an existing pending SlotExecutionVertexAssignment, which will otherwise cause slot leaks.

Reply:
> We can add a check in allocateSlotsAndDeploy() to guarantee that DefaultExecutionSlotAllocator has no pending slot assignment for the vertices to schedule. The check will fail only if there is a bug.

That would be good to have.

> It should not happen because the cancelFuture returned by cancelExecutionVertex completes only if executionSlotAllocator.cancel() completes.
And a vertex will only be restarted if its cancelFuture completes.

That sounds right. I still don't like the use of `whenComplete` here. If the slot is already provisioned, then `DefaultExecutionSlotAllocator#cancel()` should be a no-op. If the slot future is not completed, then it also cannot be assigned to the execution. I don't see harm to cancel the slot allocation independently. 

I think the difference to `Execution#allocateAndAssignSlotForExecution()`, is that DefaultScheduler#cancelExecutionVertex() is not an asynchronous callback. If `cancelExecutionVertex()` is invoked, we already know that the vertex must be cancelled. Using `whenComplete` implies that the `executionSlotAllocator.cancel()` must be invoked after the vertex cancellation is finished. If there is no such dependency, I'd prefer to keep the code a bit simpler by removing `whenComplete`. Of course, if there is such a dependency, let's keep it.

Reply:
> If the slot is already provisioned, then DefaultExecutionSlotAllocator#cancel() should be a no-op. If the slot future is not completed, then there cannot be any slot assigned to the execution.

I think you are right. Will simplify it.

Reply:
done.

Reply:
Ok, the remainder of the PR looked good to me.

Reply:
The check for duplicated slot request is done in commit 4270dd87c5e6657708d32bc1dd71cd07ccb9e09f.